### PR TITLE
Feature/scrapper

### PR DESCRIPTION
**Pull request by camescopetech**

### CODE REVIEW

File: scrapper2.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+
+app = FastAPI()
+templates = Jinja2Templates(directory="templates")
+
+@app.get("/", response_class=Response)
+async def read_root(request: Request):
+    return templates.TemplateResponse("index.html", {"request": request})
+
+@app.get("/bpi", response_class=Response, responses={200: {"content": {"application/rss+xml": {}}}})
```
Comment (ldt) [inline scrapper2.py]: trop d'accolades ?

File: scrapper2.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+async def get_bpi_feed():
+    return get_rss_feed('feeds/bpi_feed.xml')
+
+@app.get("/gnius", response_class=Response, responses={200: {"content": {"application/rss+xml": {}}}})
```
Comment (ldt) [inline scrapper2.py]: idem, trop d'accolades ?

File: README.md
Comment (ldt) [inline README.md:1]: need to explain that generate_feeds.sh should be scheduled every hour and provide an example of the cron config line

File: generate_feeds.sh [hunk 3]
```diff
@@ -1,7 +1,5 @@
 #!/bin/bash
 
-python3 -m venv .venv
-
 source .venv/bin/activate
 
 pip install --upgrade pip
```
Comment (ldt) [inline generate_feeds.sh]: why?

Comment (ldt) [reply generate_feeds.sh]: also, why `pip install -r requirements.txt` ?

File: generate_feeds.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+from scrappers.gnius import GniusScrapper
+from scrappers.bpi import BpiScrapper
+import os
+
+def main():
+    feeds_dir = 'feeds'
```
Comment (ldt) [inline generate_feeds.py]: beware if the script is launched from a different folder than the current one.
You should get the file's folder and use it as the root folder

File: requirements.txt [hunk 5]
```diff
@@ -2,4 +2,7 @@ feedgen>=0.9.0
 requests>=2.24.0
 beautifulsoup4==4.9.3
 sentry-sdk==1.30.0
-dateparser==1.2.0
\ No newline at end of file
+dateparser==1.2.0
+feedparser==6.0.11
+fastapi==0.111.0
+pytest==8.2.2
```
Comment (ldt) [inline requirements.txt]: pytest should not be installed on prod servers.
That's why it is already in a second `requirements-dev.txt` file.

File: scrapper.py [hunk 6]
```diff
@@ -1,40 +1,32 @@
...
 
-    def do_GET(self):
-        self._set_headers()
-        rss = '<?xml version="1.0" encoding="UTF-8"?><xml></xml>'.encode(
-            'utf8')
-        if (self.path[0:4] == "/bpi"):
-            rss = self.bpi_scrapper.generate_feed(verbose=False)
-        elif (self.path[0:6] == "/gnius"):
-            rss = self.gnius_scrapper.generate_feed(verbose=False)
+@app.get("/bpi", response_class=Response, responses={200: {"content": {"application/rss+xml": {}}}})
```
Comment (ldt) [inline scrapper.py]: too many {} ?

File: scrapper.py [hunk 7]
```diff
@@ -1,40 +1,32 @@
...
-
-def main(server_class=HTTPServer, handler_class=RssRequestHandler, addr="localhost", port=8000):
-    handler_class.bpi_scrapper = BpiScrapper()
-    handler_class.gnius_scrapper = GniusScrapper()
-    server_address = (addr, port)
-    httpd = server_class(server_address, handler_class)
-
-    print(f"Starting httpd server on {addr}:{port}")
-    httpd.serve_forever()
+@app.get("/gnius", response_class=Response, responses={200: {"content": {"application/rss+xml": {}}}})
```
Comment (ldt) [inline scrapper.py]: too many {} ?

File: scrappers/gnius.py [hunk 8]
```diff
@@ -32,33 +43,39 @@ def scrapPage(self, pageNumber, verbose=False):
...
-            count_for_current_page = len(posts_on_current_page)
-            page = page + 1
-        return posts
+    def print_data(self, verbose=False):
+        posts = self.scrapPages()
+        pp = pprint.PrettyPrinter(indent=4)
+        pp.pprint(posts)
+        print(f"{len(posts)} extracted")
+
+    def write_feed_to_file(self, feed, filename):
```
Comment (ldt) [inline scrappers/gnius.py]: could be factorized with the same code from the BPI scrapper

File: scrappers/gnius.py [hunk 9]
```diff
@@ -32,33 +43,39 @@ def scrapPage(self, pageNumber, verbose=False):
...
+        with open(filename, 'wb') as file:
+            file.write(feed)
+
+    def get_full_article_content(self, article_url):
```
Comment (ldt) [inline scrappers/gnius.py]: could be factorized with almost the same code from the BPI scrapper

File: tests/test_bpi_scrapping.py [hunk 10]
```diff
@@ -39,6 +39,16 @@ def test_get_article_title_link(bpi_data_soup, bpi_scrapper):
         "https://www.bpifrance.fr/nos-appels-a-projets-concours/appel-a-projets-france-chili-greentech",
     )
 
+def test_get_article_link(bpi_data_soup, bpi_scrapper):
```
Comment (ldt) [inline tests/test_bpi_scrapping.py]: why this test? What is being tested that was not in the other test?

File: tests/test_bpi_scrapping.py [hunk 11]
```diff
@@ -39,6 +39,16 @@ def test_get_article_title_link(bpi_data_soup, bpi_scrapper):
...
+    articles = bpi_scrapper.get_articles(bpi_data_soup)
+    _, link = bpi_scrapper.get_article_title_and_link(articles[0])
+    assert link == "https://www.bpifrance.fr/nos-appels-a-projets-concours/appel-a-projets-france-chili-greentech"
+
+def test_get_article_full_content(bpi_data_soup, bpi_scrapper):
```
Comment (ldt) [inline tests/test_bpi_scrapping.py:42]: how do you test nominal cases, edge cases?
ie what should happen if the article link leads to a 404 (if the article has been removed by BPI)?

Comment (camescopetech) [reply tests/test_bpi_scrapping.py:42]: @ldt does testing the error is thrown and well executed ok ?
